### PR TITLE
debt - introduce `isUriComponents` and adopt in a few untrusted places

### PR DESCRIPTION
**Pull request by bpasero**
cc @jrieken

### CODE REVIEW

**Review by aeschli — APPROVED** [review]

File: src/vs/base/test/common/uri.test.ts [hunk 1]
```diff
@@ -602,4 +602,15 @@ suite('URI', () => {
 		assert.strictEqual(URI.parse('http://user@[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:80/index.html').toString(), 'http://user@[fedc:ba98:7654:3210:fedc:ba98:7654:3210]:80/index.html');
 		assert.strictEqual(URI.parse('http://us[er@[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:80/index.html').toString(), 'http://us%5Ber@[fedc:ba98:7654:3210:fedc:ba98:7654:3210]:80/index.html');
 	});
+
+	test('isUriComponents', function () {
+		assert.strictEqual(isUriComponents(URI.parse('file:///c:/bazz')), true);
+		assert.strictEqual(isUriComponents({ scheme: 'file', authority: 'server', path: '/shares/c#f', query: 'q', fragment: 'f' }), true);
+		assert.strictEqual(isUriComponents({ scheme: '', authority: '', path: '', query: '', fragment: '' }), true);
+		assert.strictEqual(isUriComponents({ scheme: '', authority: '', path: '' }), false);
```
Comment (jrieken) [inline src/vs/base/test/common/uri.test.ts:610]: @bpasero I believe `URI.toJSON` uses `undefined` as default, e.g `URI.parse('file:///c:/bazz').toJSON()` has `query` and `fragment` set to `undefined`

Comment (bpasero) [reply src/vs/base/test/common/uri.test.ts:610]: @jrieken the method `isUriComponents` is strictly asserting the shape of the `UriComponents` interface:

https://github.com/microsoft/vscode/blob/09dde1e345aa7278ee5c82f926261f128d4d965e/src/vs/base/common/uri.ts#L401-L407

Shouldn't we then change the interface to set `query` and `fragment` as `string | undefined`?

Comment (bpasero) [reply src/vs/base/test/common/uri.test.ts:610]: Looks like there is a cast to `UriState` here which hides the issue:

https://github.com/microsoft/vscode/blob/09dde1e345aa7278ee5c82f926261f128d4d965e/src/vs/base/common/uri.ts#L457-L459

Comment (jrieken) [reply src/vs/base/test/common/uri.test.ts:610]: Yeah, there seems to be a typing issue. I believe everything can be undefined except for `scheme`, at least that's how the URI works, you can have `file` or `file?query` etc pp

**Comment by bpasero** [comment]
I am not so sure anymore this change is good and would not break the places where we check for URI components today. We can make everything optional except for `scheme` in `UriComponents` and that seems to compile fine, but it turns out that not even `scheme` is enforced:

```ts
const u = URI.revive({ scheme: '', path: '', query: '', fragment: '', authority: ''});
console.log(u.toJSON());
```

This will just print `{$mid: 1}`.

**Pull request closed without merging**